test/common/http/conn_manager_impl_test.cc

```diff
@@ -2064,6 +2064,175 @@ TEST_F(HttpConnectionManagerImplTest, TestStreamIdleAccessLog) {
+
+  // Clear and refresh the route cache (checking clusterInfo refreshes the route cache)
+  decoder_filters_[0]->callbacks_->clearRouteCache();
+  decoder_filters_[0]->callbacks_->clusterInfo();
+
+  Event::MockTimer* timer = setUpTimer();
+
+  // Set a max duration of 30ms and make sure a 30ms timer is set.
+  {
+    EXPECT_CALL(*timer, enableTimer(std::chrono::milliseconds(30), _));
+    EXPECT_CALL(route_config_provider_.route_config_->route_->route_entry_, maxStreamDuration())
+        .Times(2)
+        .WillRepeatedly(Return(std::chrono::milliseconds(30)));
+    decoder_filters_[0]->callbacks_->clearRouteCache();
+    decoder_filters_[0]->callbacks_->clusterInfo();
+  }
+
+  // Clear the timeout and make sure the timer is disabled.
+  {
+    EXPECT_CALL(*timer, disableTimer());
+    EXPECT_CALL(route_config_provider_.route_config_->route_->route_entry_, maxStreamDuration())
+        .Times(1)
+        .WillRepeatedly(Return(absl::nullopt));
+    decoder_filters_[0]->callbacks_->clearRouteCache();
+    decoder_filters_[0]->callbacks_->clusterInfo();
+  }
+
+  // Add a gRPC header, but not a gRPC timeout and verify the timer is unchanged.
+  latched_headers->setGrpcTimeout("1M");
+  {
+    EXPECT_CALL(*timer, disableTimer());
+    EXPECT_CALL(route_config_provider_.route_config_->route_->route_entry_, maxStreamDuration())
+        .Times(1)
+        .WillRepeatedly(Return(absl::nullopt));
+    decoder_filters_[0]->callbacks_->clearRouteCache();
+    decoder_filters_[0]->callbacks_->clusterInfo();
+  }
+
+  // With a gRPC header of 1M and a gRPC header max of 0, respect the gRPC header.
+  {
+    EXPECT_CALL(route_config_provider_.route_config_->route_->route_entry_, grpcTimeoutHeaderMax())
+        .Times(AnyNumber())
+        .WillRepeatedly(Return(std::chrono::milliseconds(0)));
+    EXPECT_CALL(*timer, enableTimer(std::chrono::milliseconds(60000), _));
+    decoder_filters_[0]->callbacks_->clearRouteCache();
+    decoder_filters_[0]->callbacks_->clusterInfo();
+  }
+
+  // With a gRPC header and a larger gRPC header cap, respect the gRPC header.
+  {
+    EXPECT_CALL(route_config_provider_.route_config_->route_->route_entry_, grpcTimeoutHeaderMax())
+        .Times(AnyNumber())
+        .WillRepeatedly(Return(std::chrono::milliseconds(20000000)));
+    EXPECT_CALL(*timer, enableTimer(std::chrono::milliseconds(60000), _));
+    decoder_filters_[0]->callbacks_->clearRouteCache();
+    decoder_filters_[0]->callbacks_->clusterInfo();
+  }
+
+  // With a gRPC header and a small gRPC header cap, use the cap.
+  {
+    EXPECT_CALL(route_config_provider_.route_config_->route_->route_entry_, grpcTimeoutHeaderMax())
+        .Times(AnyNumber())
+        .WillRepeatedly(Return(std::chrono::milliseconds(20)));
+    EXPECT_CALL(*timer, enableTimer(std::chrono::milliseconds(20), _));
+    decoder_filters_[0]->callbacks_->clearRouteCache();
+    decoder_filters_[0]->callbacks_->clusterInfo();
+  }
+
+  latched_headers->setGrpcTimeout("0m");
```

Comment:
If there's no deadline set on the client, this header won't be present at all.  I don't think it will ever be set to 0.